### PR TITLE
Add support for probability/thresholds in meanIoU

### LINKED ISSUE #39173: tf.keras.metrics.MeanIoU API is practically unusable without a threshold
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip3
- TensorFlow version (use command below): 2.1.0
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**

tf.keras.metrics.MeanIoU's constructor implementation does not take a threshold or list of thresholds as input argument. This is not only inconsistent the API used by other metrics (e.g. tf.keras.metrics.TruePositives, tf.keras.metrics.FalseNegatives) but also renders the API practically unusable because the outputs (i.e. predictions) from a network would generally be probability values in range from 0 to 1 and not a perfect 0 or 1 values. Hence, unless the constructor takes thresholds as argument and applies it to predictions before computing IoU, it is practically useless. It would always end up showing 0.5, or 0.25 or whatever the baseline random guess IOU happens to be in a given problem. 

**Describe the expected behavior**

tf.keras.metrics.MeanIoU constructor should take threshold values as input and also apply those before computing the IoU.

**Standalone code to reproduce the issue**
None required because the docs https://www.tensorflow.org/api_docs/python/tf/keras/metrics/MeanIoU proves the point where it only shows a example where preds are already binary values. 

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

### PR DESCRIPTION
**Pull request by TylerADavis**
Fixes #39173

Currently the meanIoU function requires that if there are `num_classes` in the model's output, each value in `y_pred` must be an integer in the range `[0, num_classes-1]`. This is a problem because most ML models will generate outputs of `num_classes` probabilities in the range `[0-1]` (shape `[dim1, dim2, ..., num_classes]` or `[dim1, dim2, ..., num_classes -1 ]`) , not outputs of shape `[dim1, dim2, ...]` with integer class values. #39173 discusses this further. 

Adding support for probabilities and thresholding will require changes to the API, and I would love to hear your thoughts on the best way to accomplish this.

## This PR adds the following:
* New optional `threshold` argument in the `meanIoU` initializer. Single configurable threshold shared across all predicted classes, used to turn 0-1 probabilities into 0 or 1 integer predictions
* Support for passing in probabilities instead of integer class labels
* New tests

## TODO:
* Add some sort of toggle to select between `classID` mode and `probability` mode. This is needed because if there are more than two classes and `y_pred` is in the current format, values will include integers `[0, 1, 2, ... n]`, and the probability mode would turn all values over the threshold to `1`, giving incorrect results
    * What is the best way to do this? One option would just be to add a `from_probability` field to `meanIoU` that defaults to false, but I'm not sure if there's a better way. #45690 is working towards adding support for logits, so we may need a way to switch between class labels, probabilities, and logins.
* Ensure that results are correct when `num_classes > 2` (no tests at present)
* When in probability mode, ensure that the last dimension of `y_pred` and `y_true` is either `num_classes` or `num_classes - 1`

### CODE REVIEW

File: tensorflow/python/keras/metrics_test.py [hunk 1]
```diff
@@ -1360,6 +1360,28 @@ def test_multi_dim_input(self):
...
+    result = m_obj(y_true, y_pred)
+
+    # y_pred gets thresholded to [[0, 1], [1, 1]]
+    # cm = [[1, 1],
+    #       [0, 2]]
+    # sum_row = [2, 2], sum_col = [1, 3], true_positives = [1, 2]
+    # iou = true_positives / (sum_row + sum_col - true_positives))
+
+    # 2 / 3
+    # 1 / 2
```
Comment (TylerADavis) [inline tensorflow/python/keras/metrics_test.py:1381]: Delete

**Comment by TylerADavis** [comment]
As an aside, are there any guides for debugging TensorFlow Python unit tests under Bazel?

I saw from [this stack overflow post](https://stackoverflow.com/questions/51709590/can-i-use-python-debugger-in-bazel-test) that you can do something like `bazel run --run_under="/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/pdb.py"
`, but I'm not sure if there is a better way to accomplish this. The workflow I'd like is to be able to run just one test in `metrics_test.py` at a time (saving time on tests I don't care about) and step through it with a debugger.

In the span of working on this PR I was just using print statements in combination with `--test_output=streamed`, and I'm assuming there must be a better way to do this.

**Comment by motionlife** [comment]
I would prefer, [#45690](https://github.com/tensorflow/tensorflow/pull/45690), which doesn't change the original interface, and dealing with both situation of `y_pred` being passed as logits tensor or probability tensor. Also, it supports multi-class label rather than just binary classification.

**Comment by TylerADavis** [comment]
Thanks for your feedback @motionlife ! I took another look at your PR, and I do think it better addresses the core of what I was trying to get to with this PR.

As far as I can tell, the only situations in which `greater_equal()` will be more helpful than `argmax()` are:
1) When users want to be able to adjust the threshold themselves
2) When one element in the output may have multiple labels, e.g. this pixel is both a person *and* a t-shirt. (I'm not sure how common this is)


For now #45690 meets my needs, but if there is interest in addressing the two points above for #39173 I would be glad to contribute.

**Comment by deeb02** [comment]
Hi @TylerADavis, Thank you for the contribution.
For now we also think #45690 is a better approach because it requires no API change. You are right that it doesn't support multi-label classification, which is a bummer, but as far as I can see, your approach (as it is right now) doesn't either, because it still uses confusion_matrix which AFAIK assumes a single class per element. A multi-label MeanIoU would also be a valuable addition IMO.

**Review by deeb02 — COMMENTED** [review]
Favoring a non API-altering solution for now.

**Comment by deeb02** [comment]
Hi @TylerADavis,
As I said before, we preferred a non PI changing solution for the moment. But if your change adds support to multi-label classification, that's good justification for changing the API. Fee free to submit this PR again, addressing the TODOs you yourself placed.
